crypto/ts/ts_rsp_verify.c

```diff
@@ -198,52 +198,58 @@ static int ts_check_signing_certs(PKCS7_SIGNER_INFO *si,
                                   STACK_OF(X509) *chain)
 {
     ESS_SIGNING_CERT *ss = ossl_ess_signing_cert_get(si);
-    STACK_OF(ESS_CERT_ID) *cert_ids = NULL;
     ESS_SIGNING_CERT_V2 *ssv2 = ossl_ess_signing_cert_v2_get(si);
-    STACK_OF(ESS_CERT_ID_V2) *cert_ids_v2 = NULL;
     X509 *cert;
     int i = 0;
     int ret = 0;
+    int id_index;
+    int chain_length;
+    int verification_length;
+    void *cert_ids_generic;
+    int (*find_cert_generic)(const void *, const X509 *);
 
     if (ss != NULL) {
-        cert_ids = ss->cert_ids;
-        cert = sk_X509_value(chain, 0);
-        if (ossl_ess_find_cert(cert_ids, cert) != 0)
-            goto err;
-
-        /*
-         * Check the other certificates of the chain if there are more than one
-         * certificate ids in cert_ids.
-         */
-        if (sk_ESS_CERT_ID_num(cert_ids) > 1) {
-            for (i = 1; i < sk_X509_num(chain); ++i) {
-                cert = sk_X509_value(chain, i);
-                if (ossl_ess_find_cert(cert_ids, cert) < 0)
-                    goto err;
-            }
-        }
+        cert_ids_generic = ss->cert_ids;
+        find_cert_generic = (int (*)(const void *, const X509 *))&ossl_ess_find_cert;
+        verification_length = sk_ESS_CERT_ID_num(cert_ids_generic);
     } else if (ssv2 != NULL) {
-        cert_ids_v2 = ssv2->cert_ids;
-        cert = sk_X509_value(chain, 0);
-        if (ossl_ess_find_cert_v2(cert_ids_v2, cert) != 0)
-            goto err;
-
-        /*
-         * Check the other certificates of the chain if there are more than one
-         * certificate ids in cert_ids.
-         */
-        if (sk_ESS_CERT_ID_V2_num(cert_ids_v2) > 1) {
-            for (i = 1; i < sk_X509_num(chain); ++i) {
-                cert = sk_X509_value(chain, i);
-                if (ossl_ess_find_cert_v2(cert_ids_v2, cert) < 0)
-                    goto err;
-            }
-        }
+        cert_ids_generic = ssv2->cert_ids;
+        find_cert_generic = (int (*)(const void *, const X509 *))&ossl_ess_find_cert_v2;
+        verification_length = sk_ESS_CERT_ID_V2_num(cert_ids_generic);
     } else {
         goto err;
     }
 
+    if (verification_length < 1) /* no signer's ESSCertID */
+        goto err;
+
+    chain_length = sk_X509_num(chain);
+    if (chain_length < 1) /* empty chain */
+        goto err;
+
+    /* RFC 2634, section 5.4 */
+    if (verification_length > 1) {
+        if (chain_length > 1)
+            /* validate every certificate except root */
+            verification_length = chain_length - 1;
+        else
+            /* only validate the signing certificate */
+            verification_length = 1;
+    }
```

Comment:
No matter which of the two currently competing interpretations will be the outcome of the conceptual discussion, this proposed implementation adaptation is wrong:

Here the length of the candidate validation chain (`chain_length`) governs how much of the list of required cert IDs (`verification_length`), not the other way round!

Suppose the list of cert IDs given by the signer is longer than the validation chain built up and to be checked by the verifier, i.e., `verification_length` was greater than `chain_length`.
The above code then shortens `verification_length` and thus makes the verifiier NEGLECT the remaining required cert IDs!
This cannot be correct (intended by the RFCs).

Conversely, suppose `verification_length` was at least two items smaller than `chain_length`.
in this case `verification_length` becomes too large, namely `chain_length - 1`
and thus in the below loop `cert = sk_X509_value(chain, i)` will
go wrong. The function will not crash just because `ossl_ess_find_cert()` is defensive on `cert == NULL`.

@t8m, therefore you should retract your approval.

Reply:
> Conversely, suppose `verification_length` was at least two items smaller than `chain_length`.
> in this case `verification_length` becomes too large, namely `chain_length - 1`
> and thus in the below loop `cert = sk_X509_value(chain, i)` will
> go wrong. The function will not crash just because `ossl_ess_find_cert()` is defensive on `cert == NULL`.

I do not think that is correct interpretation of the code. The cert will not be NULL as you're always within the chain_length - 1.

Reply:
I'm leaving my approval there until the discussion is resolved.

Reply:
I'm leaving my approval there until the discussion is resolved.

> Suppose the list of cert IDs given by the signer is longer than the validation chain built up and to be checked by the verifier, i.e., `verification_length` was greater than `chain_length`.
> The above code then shortens `verification_length` and thus makes the verifiier NEGLECT the remaining required cert IDs!
> This cannot be correct (intended by the RFCs).

For this my answer is that the original code does not prevent this either.

Reply:
> I do not think that is correct interpretation of the code.
> The cert will not be NULL as you're always within the chain_length - 1.

Ah, right, although in the following loop `i` somewhat confusingly seems to range over the cert ID list, in fact it ranges over the chain except its last element:
```
for (i = 0; i < verification_length; i++) {
        cert = sk_X509_value(chain, i);
```
so my conclusion on `cert == NULL` for this case was wrong.

Reply:
> For this my answer is that the original code does not prevent this either.

Yeah, the original code was even more wrong :frowning_face: